slsim/Sources/SourceVariability/variability.py

```diff
@@ -36,4 +77,346 @@ def variability_at_time(self, observation_times):
+            if response_function_amplitudes is not None:
+                raise ValueError(
+                    "Please provide only a wavelength or only a response function. Not both!"
+                )
+            response_function = self.define_new_response_function(
+                rest_frame_wavelength_in_nanometers
+            )
+            length_in_days = int(len(response_function) * gravitational_radius_in_days)
+            time_lag_axis = np.linspace(0, length_in_days, len(response_function))
+        if rest_frame_wavelength_in_nanometers is None:
+            if response_function_amplitudes is None:
+                raise ValueError("Please provide a wavelength or a response function.")
+
+            response_function = response_function_amplitudes
+
+            if response_function_time_lags is not None:
+                if len(response_function_time_lags) != len(
+                    response_function_amplitudes
+                ):
+                    raise ValueError(
+                        "The time lag array and response function array must match in length."
+                    )
+
+                time_lag_axis = response_function_time_lags
+                length_in_days = int(time_lag_axis[-1] - time_lag_axis[0])
+            else:
+                length_in_days = int(
+                    len(response_function) * gravitational_radius_in_days
+                )
+                time_lag_axis = np.linspace(0, length_in_days, len(response_function))
+
+        interpolation_of_response_function = scipy.interpolate.interp1d(
+            time_lag_axis, response_function, bounds_error=False, fill_value=0
+        )
+
+        tau_axis = np.linspace(0, int(time_lag_axis[-1]), int(time_lag_axis[-1]))
+
+        interpolated_response_function = interpolation_of_response_function(tau_axis)
+
+        interpolation_of_signal = scipy.interpolate.interp1d(
+            self.timestamps,
+            self.signal,
+            bounds_error=False,
+            fill_value=(self.signal[0], self.signal[-1]),
+        )
+
+        signal_time_axis = np.linspace(
+            self.timestamps[0],
+            self.timestamps[-1],
+            int(self.timestamps[-1] - self.timestamps[0]),
+        )
+
+        interpolated_signal = interpolation_of_signal(signal_time_axis)
+
+        reprocessed_signal = scipy.signal.convolve(
+            interpolated_signal, (interpolated_response_function), mode="full"
+        ) / np.sum(interpolated_response_function)
+
+        return reprocessed_signal[: 1 - length_in_days]
+
+
+def sinusoidal_variability(t, amp, freq):
+    """Calculate the sinusoidal variability for a given observation time.
+
+    :param t: observation time in [days].
+    :param kwargs_model: dictionary of variability parameter associated with a source.
+    :return: variability for the given time
+    """
+    return amp * np.sin(2 * np.pi * freq * t)
+
+
+def broken_power_law(
```

Comment:
You can move this function, generic_psd() function, and lamppost_model() function  to a separate file and call here from that file.